timely/src/dataflow/operators/to_stream.rs

```diff
@@ -56,3 +60,82 @@ impl<T: Timestamp, I: IntoIterator+'static> ToStream<T, I::Item> for I where I::
         })
     }
 }
+
+/// Data and progress events of the native stream.
+pub enum Event<F: IntoIterator, D> {
+    /// Indicates that timestamps have advanced to frontier F
+    Progress(F),
+    /// Indicates that event D happened at time T
+    Message(F::Item, D),
+}
+
+/// Converts to a timely `Stream`.
+pub trait ToStreamAsync<T: Timestamp, D: Data> {
+    /// Converts a [native `Stream`](futures_util::stream::Stream) of [`Event`s](Event) into a [timely
+    /// `Stream`](crate::dataflow::Stream).
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use futures_util::stream;
+    ///
+    /// use timely::dataflow::operators::{Capture, Event, ToStream, ToStreamAsync};
+    /// use timely::dataflow::operators::capture::Extract;
+    ///
+    /// let native_stream = stream::iter(vec![
+    ///     Event::Message(0, 0),
+    ///     Event::Message(0, 1),
+    ///     Event::Message(0, 2),
+    ///     Event::Progress(Some(0)),
+    /// ]);
+    ///
+    /// let native_stream = Box::pin(native_stream);
+    ///
+    /// let (data1, data2) = timely::example(|scope| {
+    ///     let data1 = native_stream.to_stream(scope).capture();
+    ///     let data2 = vec![0,1,2].to_stream(scope).capture();
+    ///
+    ///     (data1, data2)
+    /// });
+    ///
+    /// assert_eq!(data1.extract(), data2.extract());
+    /// ```
+    fn to_stream<S: Scope<Timestamp = T>>(self: Pin<Box<Self>>, scope: &S) -> Stream<S, D>;
+}
+
+impl<T, D, F, I> ToStreamAsync<T, D> for I
+where
+    D: Data,
+    T: Timestamp,
+    F: IntoIterator<Item = T>,
+    I: futures_util::stream::Stream<Item = Event<F, D>> + ?Sized + 'static,
+{
+    fn to_stream<S: Scope<Timestamp = T>>(mut self: Pin<Box<Self>>, scope: &S) -> Stream<S, D> {
+        source(scope, "ToStreamAsync", move |capability, info| {
+            let activator = Arc::new(scope.sync_activator_for(&info.address[..]));
+
+            let mut cap_set = CapabilitySet::from_elem(capability);
+
+            move |output| {
+                let waker = futures_util::task::waker_ref(&activator);
+                let mut context = Context::from_waker(&waker);
+
+                // Consume all the ready items of the source_stream and issue them to the operator
+                while let Poll::Ready(item) = self.as_mut().poll_next(&mut context) {
+                    match item {
+                        Some(Event::Progress(time)) => {
+                            cap_set.downgrade(time);
+                        }
+                        Some(Event::Message(time, data)) => {
+                            output.session(&cap_set.delayed(&time)).give(data);
```

Comment:
Minor, but this will eventually have throughput issues. At least, timely can send data around much faster than it can mint capabilities. It may eventually be smart to have the `D` of the enum implement `IntoIterator<Item = D2>`, though perhaps .. that is just a flatmap after this. Idk. Doesn't block anything here, but wanted to call out!